tests/test_qutrit_device.py

```diff
@@ -0,0 +1,1022 @@
+            dev.obs_queue
+
+    def test_unsupported_observables_raise_error(self, mock_qutrit_device):
+        """Tests that the operations are properly applied and queued"""
+        U = unitary_group.rvs(3, random_state=10)
+
+        with qml.tape.QuantumTape() as tape:
+            queue = [qml.QutritUnitary(U, wires=0)]
+            observables = [qml.expval(qml.Hadamard(0))]
+
+        with pytest.raises(DeviceError, match="Observable Hadamard not supported on device"):
+            dev = mock_qutrit_device()
+            dev.execute(tape)
+
+    def test_unsupported_observable_return_type_raise_error(self, mock_qutrit_device, monkeypatch):
+        """Check that an error is raised if the return type of an observable is unsupported"""
+        U = unitary_group.rvs(3, random_state=10)
+
+        with qml.tape.QuantumTape() as tape:
+            qml.QutritUnitary(U, wires=0)
+            qml.measurements.MeasurementProcess(
+                return_type="SomeUnsupportedReturnType", obs=qml.Identity(0)
+            )
+
+        with monkeypatch.context() as m:
+            m.setattr(QutritDevice, "apply", lambda self, x, **kwargs: None)
+            with pytest.raises(
+                qml.QuantumFunctionError, match="Unsupported return type specified for observable"
+            ):
+                dev = mock_qutrit_device()
+                dev.execute(tape)
+
+
+class TestParameters:
+    """Test for checking device parameter mappings"""
+
+    def test_parameters_accessed_outside_execution_context(self, mock_qutrit_device):
+        """Tests that a call to parameters outside the execution context raises the correct error"""
+
+        with pytest.raises(
+            ValueError,
+            match="Cannot access the free parameter mapping outside of the execution context!",
+        ):
+            dev = mock_qutrit_device()
+            dev.parameters
+
+
+class TestExtractStatistics:
+    @pytest.mark.parametrize("returntype", [Expectation, Variance, Sample, Probability, State])
+    def test_results_created(self, mock_qutrit_device_extract_stats, monkeypatch, returntype):
+        """Tests that the statistics method simply builds a results list without any side-effects"""
+
+        class SomeObservable(qml.operation.Observable):
+            num_wires = 1
+            return_type = returntype
+
+        obs = SomeObservable(wires=0)
+
+        with monkeypatch.context() as m:
+            dev = mock_qutrit_device_extract_stats()
+            results = dev.statistics([obs])
+
+        assert results == [0]
+
+    def test_results_no_state(self, mock_qutrit_device_extract_stats, monkeypatch):
+        """Tests that the statistics method raises an AttributeError when a State return type is
+        requested when QutritDevice does not have a state attribute"""
+        with monkeypatch.context():
+            dev = mock_qutrit_device_extract_stats()
+            delattr(dev.__class__, "state")
+            del QubitDevice.state
```

Comment:
This was the reason that `state` needed to be redefined in the `QutritDevice` class, right? Maybe provide some details of that here?

Reply:
Yes, this was why.